### ISSUE TITLE
Poor error message while passing generic type argument to a method as a parameter

### ISSUE THREAD

**Issue by drbcode** [dart-lang/sdk#36936]
code: a.dart
```dart
class Hello{
  void world<T>(T a) {}
  void world2<T>(T a) {}
}

class Bye {
  int a = 1;
}

class Something extends Hello {
  void world<Bye>(Bye a) {
    printBye(a);
  }

  void world2<int>(int a) {
    printInt(a);
  }
}

void printBye(Bye b) {
  print(b.a);
}

void printInt(int b) {
  print(b);
}
```

Error messages:
![Screenshot_2019-05-12_05-53-53](https://user-images.githubusercontent.com/41248528/57576317-61f83780-747a-11e9-9bdc-881b4ccabd7e.png)

Above error messages can be improved.

Dart sdk version: 2.3.0

**Comment by mraleph** [comment]
The error is correct, though somewhat confusingly formatted (fyi @stereotype441).

In the signature `world2<int>(int a)` `int` becomes _a type parameter_, it does not refer to the actual type `int`. In other words: `world2<int>(int a)` is not different in any way from `world2<X>(X a)`.

Also note that you can't override a method with type arguments with a method without type arguments. (and in general overriding `world<T>(T x)` with `world(Bye x)` violates [Liskov's substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)).

**Comment by drbcode** [comment]
If I can't override then why the analyzer did not show any error? 

That's the issue, analyzer is not showing helpful messages.

**Comment by mraleph** [comment]
@drbcode you can't overwrite with `world(Bye x)`, however you can overwrite with `world<Bye>(Bye x)` - which is exactly the same as `world<X>(X x)` and not at all related to `world(Bye x)`

**Comment by drbcode** [comment]
Yes, I did just that but the analyzer still reports they are the same types.

Messages are confusing:
```
The argument type 'Bye ()' can't be assigned to the parameter type 'Bye ()'
The argument type 'int ()' can't be assigned to the parameter type 'int ()'
```

Using cast operator is also confusing:

![cast](https://user-images.githubusercontent.com/41248528/57579392-bae6c080-74b8-11e9-9005-19135b8221ce.png)

This is not the case with generic classes where everything works as expected.

**Comment by mraleph** [comment]
@drbcode let me repeat again: in code `world<Bye>(Bye x)` `Bye` does not refer to the type `Bye`. It refers to the type parameter `Bye` which you declared. For all intents and purposes you could also write the same code as `world<X>(X x)` or `world<T>(T x)` or `world<Z>(Z x)` - all of these mean exactly the same thing. 

Here is what you essentially wrote (I am using name `T` instead of using name `Bye` for the type parameter - which confuses you):

```
class Something extends Hello {
  void world<T>(T x) {
    printBye(x as T);
  }
}
```

Here you should see why analyzer give you the warning it gives. 

You seem to think that writing `world<Bye>(Bye x)` creates some sort of specialization of generic method with `T` replaced by `Bye`. This is not what it means in Dart.

**Comment by drbcode** [comment]
This clears the confusion.
Thank you!

**Comment by stereotype441** [comment]
> The error is correct, though somewhat confusingly formatted (fyi @stereotype441).

CC @bwilkerson, who has been thinking about how to improve error messages from the analyzer.

**Comment by bwilkerson** [comment]
I've updated the title, and am re-opening so that this issue doesn't get lost.

I think we should
- reword the message to make it more clear that the two types are different even though they have the same name, and
- add context messages indicating the places at which the two types are defined.
